src/connection.ts

```diff
@@ -899,44 +895,21 @@ export class Connection extends EventEmitter {
       }
     }
 
-    // Stop sending if there's no more to send
-    // TODO don't stop if there's still data to send
-    if (amountToSend.isEqualTo(0) && !sendingData) {
-      this.debug(`packet value is 0 and there is no data to send so we'll send this packet and then stop`)
-      this.sending = false
-      // TODO figure out if there are control frames we need to send and stop sending if not
-    }
-
-    // TODO is this needed now that stream close frames are sent out when the stream.close event is emitted?
-    // Tell other side which streams are closed
-    if (!this.closed) {
-      for (let [_, stream] of this.streams) {
-        if (stream.isOpen() || stream._sentEnd) {
-          continue
-        }
-        if (stream._getAmountAvailableToSend().isGreaterThan(0)) {
-          this.debug(`stream ${stream.id} is closed but still has money to send, not sending end frame yet`)
-          continue
-        }
-        if (stream._hasDataToSend()) {
-          this.debug(`stream ${stream.id} is closed but still has data to send, not sending end frame yet`)
-          continue
-        }
-        const streamEndFrame = (stream._errorMessage
-          ? new StreamCloseFrame(stream.id, ErrorCode.ApplicationError, stream._errorMessage)
-          : new StreamCloseFrame(stream.id, ErrorCode.NoError, ''))
-
-        // Make sure the packet has space left
-        if (streamEndFrame.byteLength() > bytesLeftInPacket) {
-          // TODO make sure it will actually make another pass to send these later
-          this.debug('not sending more stream end frames because the packet is full')
-          break
+    // Check if we can stop sending
+    if (amountToSend.isEqualTo(0)) {
+      if (requestPacket.frames.length === 0) {
+        this.sending = false
+        return
+      } else {
+        // Check if any of the frames are Close, Data, or Money Frames
+        // If none of those frames are present then we can stop sending
+        // after the current frames in the request packet are sent.
+        if (!requestPacket.frames.find(frame =>
+            ((frame.type === FrameType.StreamClose)
```

Comment:
If the retry is the only reason to do it this way, I would just set `this.sending` to false but make the function that handles rejected packets start the send loop again

Reply:
Also, if a frame is rejected, it shouldn't stop after a single retry. What if it needs 2 or 3 to get through?

Reply:
> Also, if a frame is rejected, it shouldn't stop after a single retry. What if it needs 2 or 3 to get through?

sendPacket gets called, sends the close, then calls undo which adds the close frame back to the queued frames, since sending is not false it will try the loop again and pick it up so it will keep sending the close, money, or data frames on rejects.